File: Unit2-05/Typescript/Vehicle.ts
/**
* This class creates a vehicle
*
* By:      Kenny Le
* Version: 1.0
* Since:   2024-03-27
*/

export class Vehicle {
  // Class variables

  // Vehicle license plate number
  licensePlateNumber: string
  // Vehicle colour
  private colour: string
  // Number of doors on the vehicle
  private doors: number
  // Current speed of the vehicle
  private speed: number
  // Maximum speed of the vehicle
  private maxSpeed: number

  // Creates the base vehicle
  public constructor(startColour: string, startPlateNumber: string,startDoors: int, startMaxSpeed: number) {
    this.speed = 0
    this.maxSpeed = startMaxSpeed
    this.doors = startDoors
    this.licensePlateNumber = startPlateNumber
    this.colour = startColour
  }

  // Gets the license plate number of the vehicle
  public get licensePlateNumber(): string {
    return this.licensePlateNumber
  }

  // Gets the colour of the vehicle
  public get colour(): string {
    return this.colour
  }

  // Gets the number of doors on the vehicle
  public get doors(): number {
    return this.doors
  }

  // Gets the current speed of the vehicle
  public get speed(): number {
    return this.speed
  }

  // Gets the maximum speed of the vehicle
  public get maxSpeed(): number {
    return this.maxSpeed
  }

  // Creates a string about the status of the car
  public get status(): string {
    let statusString = ` -> Speed: ${this.speed}\n`
    statusString += ` -> MaxSpeed: ${this.maxSpeed}\n`
    statusString += ` -> Number of doors: ${this.doors}\n`
    statusString += ` -> License Plate: ${this.licensePlateNumber}\n`
    statusString += ` -> Color: ${this.colour}`

    return statusString
  }

  // Sets the colour of the vehicle
  public set colour(newColour: string) {
    this.colour = newColour
  }

  // Sets the license plate number of the vehicle
  public set licensePlateNumber(newPlateNumber: string) {
    this.licensePlateNumber = newPlateNumber
  }

  // Accelerates the car, changing the speed
  public accelerate(accelerationPower: number, accelerationTime: number) {
    if (accelerationPower > 0 && accelerationTime > 0) {
      this.speed = (accelerationPower * accelerationTime) + this.speed;
      if (this.speed > this.maxSpeed) {
        this.speed = this.maxSpeed;
      }
  } else {
      console.error("Acceleration power and time must be positive numbers.");
  }
}

  // Brakes the car, changing the speed
  public brake(brakePower: number, brakeTime: number) {
    if (brakePower > 0 && brakeTime > 0) {
      this.speed = this.speed - (brakePower * brakeTime);
      if (this.speed < 0) {
        this.speed = 0;
      }
    } else {
      console.error("Brake power and time must be positive numbers.");
    }
  }
}

File: Unit2-05/Java/Vehicle.java
/*
* This class creates a vehicle
*
* @author  Kenny Le
* @version 1.0
* @since   2024-04-12
*/

/**
 * This is the class.
*/
public class Vehicle {
    /**
     * Variable for the class.
     *
     * @param licensePlateNumber the license plate number on the vehicle
    */
    private String licensePlateNumber = "";

    /**
     * Variable for the class.
     *
     * @param colour of the vehicle
    */
    private String colour = "";

    /**
     * Variables for the class.
     *
     * @param doors number of doors on the vehicle
    */
    private int doors;

    /**
     * Variables for the class.
     *
     * @param speed of the vehicle
    */
    private double speed;

    /**
     * Variables for the class.
     *
     * @param maxSpeed of the vehicle
    */
    private double maxSpeed;

    /**
     * Creates the base vehicle.
     *
     * @param initialColour the base colour
     * @param initialPlateNumber the base license plate number
     * @param initialDoors the number of doors
     * @param initialMaxSpeed the maximum speed of the vehicle
    */
    public Vehicle(String initialColour, String initialPlateNumber,
                    int initialDoors, double initialMaxSpeed) {
        this.speed = 0;
        this.maxSpeed = initialMaxSpeed;
        this.doors = initialDoors;
        this.licensePlateNumber = initialPlateNumber;
        this.colour = initialColour;
    }

    /**
     * Gets the license plate number of the vehicle.
     *
     * @return the current licensePlateNumber
    */
    public String getLicensePlateNumber() {
        return this.licensePlateNumber;
    }

    /**
     * Gets the colour of the vehicle.
     *
     * @return the current colour
    */
    public String getColour() {
        return this.colour;
    }

    /**
     * Gets the number of doors on the vehicle.
     *
     * @return the number of doors
    */
    public int getDoors() {
        return this.doors;
    }

    /**
     * Gets the speed of the vehicle.
     *
     * @return the current speed
    */
    public double getSpeed() {
        return this.speed;
    }

    /**
     * Gets the maximum speed of the vehicle.
     *
     * @return the maximum speed
    */
    public double getMaxSpeed() {
        return this.maxSpeed;
    }

    /**
     * Creates a string about the status of the vehicle.
     *
     * @return the string about the vehicle's status
    */
    public String getStatus() {
        final String breakLine = "\n";
        String status = " -> Speed: " + this.speed + breakLine;
        status += " -> MaxSpeed: " + this.maxSpeed + breakLine;
        status += " -> Number of doors: " + this.doors + breakLine;
        status += " -> License Plate: " + this.licensePlateNumber + breakLine;
        status += " -> Color: " + this.colour + breakLine;

        return status;
    }

    /**
     * Sets the colour of the vehicle.
     *
     * @param newColour of the vehicle
    */
    public void setColour(String newColour) {
        this.colour = newColour;
    }

    /**
     * Sets the license plate of the vehicle.
     *
     * @param newPlateNumber of the vehicle.
    */
    public void setLicensePlateNumber(String newPlateNumber) {
        this.licensePlateNumber = newPlateNumber;
    }

    /**
     * Accelerates the vehicle, which changes the speed.
     *
     * @param accelerationPower power of the acceleration
     * @param accelerationTime time of the acceleration
    */
    public void accelerate(double accelerationPower, double accelerationTime) {
        if (accelerationPower > 0 && accelerationTime > 0) {
            this.speed = (accelerationPower * accelerationTime) + this.speed;
            if (this.speed > this.maxSpeed) {
                this.speed = this.maxSpeed;
            }
        } else {
            System.out.print("Acceleration power and "
                                + "time must be positive numbers.");
        }
    }

    /**
     * Brakes the vehicle, which changes speed.
     *
     * @param brakePower power of the brake
     * @param brakeTime time of the brake
    */
    public void brake(double brakePower, double brakeTime) {
        if (brakePower > 0 && brakeTime > 0) {
            this.speed = this.speed - (brakePower * brakeTime);
            if (this.speed < 0) {
                this.speed = 0;
            }
        } else {
            System.out.print("Brake power and time must be positive numbers.");
        }
    }
}
